File: leetcode-37_sudoku_solver.cpp
#include<iostream>
#include<vector>
#include<map>

using std::cout, std::vector, std::map;

void print_board(vector<vector<char>> &b) {
    for (vector<char> &l : b) {
        for (char c : l) {
            cout << c << ' ';
        }
        cout << '\n';
    }
}

class Solution {
public:
    bool isValid(vector<vector<char>>&b, int r, int c, char v) {
        for (int i=0; i<9; i++) {
            if (b[i][c] == v) return false;
            if (b[r][i] == v) return false;
            if (b[3*(r/3)+i/3][3*(c/3)+i%3] == v) return false;
        }
        return true;
    }

    bool solve(vector<vector<char>> &b) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (b[i][j] == '.') {
                    for (char ch='1'; ch<='9'; ch++) {
                        if (isValid(b, i, j, ch)) {
                            b[i][j] = ch;
                            if (solve(b)) return true;
                            b[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    void solveSudoku(vector<vector<char>>&board) {
        solve(board);
    }
};

int main() {
    Solution solution;
    vector<vector<char>> board {
        {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
        {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
        {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
        {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
        {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
        {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
        {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
        {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
        {'.', '.', '.', '.', '8', '.', '.', '7', '9'}
    };
    print_board(board);
    cout << '\n';
    solution.solveSudoku(board);
    print_board(board);
    return 0;
}

/*
37. Sudoku Solver

Write a program to solve a Sudoku puzzle by filling the empty cells.
A sudoku solution must satisfy all of the following rules:
Each of the digits 1-9 must occur exactly once in each row.
Each of the digits 1-9 must occur exactly once in each column.
Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
The '.' character indicates empty cells.


Example 1:
Input: board = 
[["5","3",".",".","7",".",".",".","."],
 ["6",".",".","1","9","5",".",".","."],
 [".","9","8",".",".",".",".","6","."],
 ["8",".",".",".","6",".",".",".","3"],
 ["4",".",".","8",".","3",".",".","1"],
 ["7",".",".",".","2",".",".",".","6"],
 [".","6",".",".",".",".","2","8","."],
 [".",".",".","4","1","9",".",".","5"],
 [".",".",".",".","8",".",".","7","9"]]

Output: 
[["5","3","4","6","7","8","9","1","2"],
 ["6","7","2","1","9","5","3","4","8"],
 ["1","9","8","3","4","2","5","6","7"],
 ["8","5","9","7","6","1","4","2","3"],
 ["4","2","6","8","5","3","7","9","1"],
 ["7","1","3","9","2","4","8","5","6"],
 ["9","6","1","5","3","7","2","8","4"],
 ["2","8","7","4","1","9","6","3","5"],
 ["3","4","5","2","8","6","1","7","9"]]

Constraints:
board.length == 9
board[i].length == 9
board[i][j] is a digit or '.'.
It is guaranteed that the input board has only one solution.
*/

File: leetcode-37_sudoku_solver.py
from typing import List

def print_board(b):
    for row in b:
        print(' '.join(row))

class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        def is_valid(r, c, ch):
            for i in range(9):
                if board[r][i] == ch:
                    return False
                if board[i][c] == ch:
                    return False
                if board[3*(r//3)+i//3][3*(c//3)+i%3] == ch:
                    return False
            return True
        
        def solve():
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        for x in ['1','2','3','4','5','6','7','8','9']:
                            if is_valid(i, j, x) == True:
                                board[i][j] = x
                                if solve() == True:
                                    return True
                                board[i][j] = '.'
                        return False
            return True
        solve()

if __name__ == '__main__':
    s = Solution()
    board =[["5", "3", ".", ".", "7", ".", ".", ".", "."],
            ["6", ".", ".", "1", "9", "5", ".", ".", "."], 
            [".", "9", "8", ".", ".", ".", ".", "6", "."], 
            ["8", ".", ".", ".", "6", ".", ".", ".", "3"], 
            ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
            ["7", ".", ".", ".", "2", ".", ".", ".", "6"], 
            [".", "6", ".", ".", ".", ".", "2", "8", "."], 
            [".", ".", ".", "4", "1", "9", ".", ".", "5"], 
            [".", ".", ".", ".", "8", ".", ".", "7", "9"]]
    print_board(board)
    print()
    s.solveSudoku(board);
    print_board(board)
